guides/testing.md

```diff
@@ -0,0 +1,163 @@
+
+Inspired by [Rspec](http://rspec.info/) it includes a domain specific language (DSL) that allows to write examples in a way similar to plain english..
+
+A basic spec looks something like this:
+
+```crystal
+require "spec"
+
+describe Array do
+  describe "#size" do
+    it "correctly reports the number of elements in the Array" do
+      [1, 2, 3].size.should eq 3
+    end
+  end
+
+  describe "#empty?" do
+    it "is empty when no elements are in the array" do
+      ([] of Int32).empty?.should be_true
+    end
+
+    it "is not empty if there are elements in the array" do
+      [1].empty?.should be_false
+    end
+  end
+end
+```
+
+## Anatomy of a spec file
+
+To use the spec module and DSL, you need to add `require "spec"` to your spec files. Many projects use a custom [spec helper](#spec-helper) which organizes these includes.
+
+In test files, specs are structured in example groups enclosed in `describe` and `context` sections. Typically a top level `describe` defines the outer unit (such as a class) to be tested by the spec. Further `describe` sections can be nested within the outer unit to specify smaller units under test (such as individual methods).
+
+For unit tests, it is recommended to follow the conventions for method names: Outer `describe` is the name of the class, inner `describe` targets methods. Instance methods are prefixed with `#`, class methods with `.`.
+
+To establish certain contexts - think *empty array* versus *array with elements* - the `context` method may be used to communicate this to the reader. It has a different name, but behaves exactly like `describe`.
+
+`describe` and `context` take a description as argument (which should usually be a string) and a block containing the individual specs or nested groupings.
+
+Concrete test cases are defined in `it` blocks. An optional (but strongly recommended) descriptive string states it's purpose and a block contains the main logic performing the test.
+
+Test cases that have been defined or outlined but are not yet expected to work can be defined using `pending` instead of `it`. They will not be run but show up in the spec report as pending.
+
+An `it` block contains an example that should invoke the code to be tested and define what is expected of it.
+
+When `spec` is included, every object has the instance methods `#should` and `#should_not`. These methods are invoked on the value being tested with an expectation as argument. If the expectation is met, code execution continues. Otherwise the example has *failed* and other code in this block will not be executed.
+
+Each example can contain multiple expectations, but it should test only one specific behaviour.
+
+## Expectations
+
+Expectations define if the value being tested (*actual*) matches a certain value or specific criteria.
+
+### Equivalence, Identity and Type
+```crystal
+actual.should eq(expected)    # passes if actual == expected
+actual.should be(expected)    # passes if actual.same?(expected)
```

Comment:
There should be a note on the difference between `==` and `same?`, and that `same?` isn't applicable on structs.